swarm/network/kademlia.go

```diff
@@ -613,187 +602,198 @@ func (k *Kademlia) string() string {
 	})
-	for _, p := range peers {
-		pk := fmt.Sprintf("%x", p)
-		if !pm[pk] {
-			log.Trace(fmt.Sprintf("%08x: known nearest neighbour %s not found", k.BaseAddr()[:4], pk[:8]))
-			return false
+
+	// iterate through nearest neighbors in the peerpot map
+	// if we can't find the neighbor in the map we created above
+	// then we don't know all our neighbors
+	// (which sadly is all too common in modern society)
+	var gots int
+	var culprits [][]byte
+	for _, p := range addrs {
+		pk := common.Bytes2Hex(p)
+		if pm[pk] {
+			gots++
+		} else {
+			log.Trace(fmt.Sprintf("%08x: known nearest neighbour %s not found", k.base, pk))
+			culprits = append(culprits, p)
 		}
 	}
-	return true
+	return gots == len(addrs), gots, culprits
 }
 
-// gotNearestNeighbours tests if all known nearest neighbours given as arguments
-// are connected peers
+// connectedNeighbours tests if all neighbours in the peerpot
+// are currently connected in the kademlia
 // It is used in Healthy function for testing only
-func (k *Kademlia) gotNearestNeighbours(peers [][]byte) (got bool, n int, missing [][]byte) {
+// TODO move to separate testing tools file
+func (k *Kademlia) connectedNeighbours(peers [][]byte) (got bool, n int, missing [][]byte) {
 	pm := make(map[string]bool)
 
+	// create a map with all peers at depth and deeper that are connected in the kademlia
+	// in order deepest to shallowest compared to the kademlia base address
+	// all bins (except self) are included (0 <= bin <= 255)
+	depth := depthForPot(k.conns, k.MinProxBinSize, k.base)
 	k.eachConn(nil, 255, func(p *Peer, po int, nn bool) bool {
-		if !nn {
+		if po < depth {
 			return false
 		}
-		pk := fmt.Sprintf("%x", p.Address())
+		pk := common.Bytes2Hex(p.Address())
 		pm[pk] = true
 		return true
 	})
+
+	// iterate through nearest neighbors in the peerpot map
+	// if we can't find the neighbor in the map we created above
+	// then we don't know all our neighbors
 	var gots int
 	var culprits [][]byte
 	for _, p := range peers {
-		pk := fmt.Sprintf("%x", p)
+		pk := common.Bytes2Hex(p)
 		if pm[pk] {
 			gots++
 		} else {
-			log.Trace(fmt.Sprintf("%08x: ExpNN: %s not found", k.BaseAddr()[:4], pk[:8]))
+			log.Trace(fmt.Sprintf("%08x: ExpNN: %s not found", k.base, pk))
 			culprits = append(culprits, p)
 		}
 	}
 	return gots == len(peers), gots, culprits
 }
 
+// connectedPotential checks whether the peer is connected to a health minimum of peers it knows about in bins that are shallower than depth
+// it returns an array of bin proximity orders for which this is not the case
+// TODO move to separate testing tools file
+func (k *Kademlia) connectedPotential() []uint8 {
+	pk := make(map[uint8]int)
+	pc := make(map[uint8]int)
+
+	// create a map with all bins that have known peers
+	// in order deepest to shallowest compared to the kademlia base address
+	depth := depthForPot(k.conns, k.MinProxBinSize, k.base)
+	k.eachAddr(nil, depth, func(_ *BzzAddr, po int, _ bool) bool {
+		pk[uint8(po)]++
```

Comment:
Probably a good idea to extract `Po` as a type, because we use it as `uint8` and `int` at different places. Or at the very least be consistent about its type.

Doesn't have to be addressed in this PR, but we shouldn't have to do conversions here.